### PR TITLE
Add support for saving new strings with multi-encoding

### LINKED ISSUE #1789: Support proper encoding of strings with multi-valued encodings
This is a spin-off of #1785.

Currently, fo-dicom correctly handles the decoding of strings with multi-valued character sets, but for the encoding only the first of the encodings is taken into account. This is only a problem for newly added strings, as strings loaded from existing data will already have the encoded byte representation saved.

Here is an example test that should pass:
```cs
        [Fact]
        public void Test_SavingPatientNameWithMultiEncoding()
        {
            var patientName = "Yamada^Tarou=山田^太郎=やまだ^たろう";
            var dataset = new DicomDataset
            {
                { DicomTag.SOPClassUID, DicomUID.SecondaryCaptureImageStorage },
                { DicomTag.SOPInstanceUID, DicomUIDGenerator.GenerateDerivedFromUUID() },
                { DicomTag.SpecificCharacterSet, "\\ISO 2022 IR 87" },
                { DicomTag.PatientName, patientName}
            };
            var dicomFile = new DicomFile(dataset);
            var stream = new MemoryStream();
            dicomFile.Save(stream);
            stream.Seek(0, SeekOrigin.Begin);
            var inFile = DicomFile.Open(stream);
            Assert.Equal(patientName, inFile.Dataset.GetString(DicomTag.PatientName));
        }
```
A few considerations taken from the other issue:
- to find errors in encodings, the used encoding shall be constructed with `EncoderFallback.ExceptionFallback` (as is already done for decoding)
- tags with VR PN shall be handled by encoding each component separately (if encoding with the first value fails)
- any encoded string part that does not use the first character set shall be surrounded by the proper escape sequences (as already used in decoding)
- we have to decide if we want to support strings or PN components which need more than one encoding to encode, as these are difficult to implement and practically non-existent

_Note_: There is already [a TODO comment](https://github.com/fo-dicom/fo-dicom/blob/66b496269f00d5ebc4903f37a2c08d8fdba2d4bc/FO-DICOM.Core/DicomElement.cs#L187) in the problematic code.

### PR DESCRIPTION
**Pull request by mrbean-bremen**
See #1789.

This implements the most relevant use case (only one encoding per value/component). 
I did not think much about optimization of the case that the first encoding cannot encode the string, as it is relatively rare. The standard case should almost not be impacted.

The problem with JIS X 201 and JIS X 208 mentioned in the issue is not handled here.

#### Checklist
- [x] The pull request branch is in sync with latest commit on the *fo-dicom/development* branch
- [x] I have updated API documentation
- [x] I have included unit tests
- [x] I have updated the change log
- [x] I am listed in the CONTRIBUTORS file

### CODE REVIEW

File: Documentation/v5/index.md [hunk 1]
```diff
@@ -29,5 +29,4 @@ Package | Description
 [fo-dicom.Imaging.Desktop](https://www.nuget.org/packages/fo-dicom.Imaging.Desktop/) | Library with referencte to System.Drawing, required for rendering into Bitmaps
 [fo-dicom.Imaging.ImageSharp](https://www.nuget.org/packages/fo-dicom.Imaging.ImageSharp/) | Library with reference to ImageSharp, can be used for platform independent rendering
 [fo-dicom.NLog](https://www.nuget.org/packages/fo-dicom.NLog/) | .NET connector to enable `fo-dicom` logging with NLog
-[fo-dicom.Codecs](https://www.nuget.org/packages/fo-dicom.Codecs/) | Cross-platform DICOM codecs for `fo-dicom`, developed by [Efferent Health] (https://github.com/Efferent-Health/fo-dicom.Codecs)
```
Comment (mrbean-bremen) [inline Documentation/v5/index.md:32]: This is unrelated - there was a space that broke that link.

File: FO-DICOM.Core/DicomDataset.cs [hunk 2]
```diff
@@ -144,7 +144,7 @@ internal void SetFallbackEncodings(Encoding[] value)
         internal Encoding[] GetEncodingsForSerialization()
         {
             return TryGetValues<string>(DicomTag.SpecificCharacterSet, out var charsets)
-                ? DicomEncoding.GetEncodings(charsets)
+                ? DicomEncoding.GetStrictEncodings(DicomEncoding.GetEncodings(charsets))
```
Comment (mrbean-bremen) [inline FO-DICOM.Core/DicomDataset.cs:147]: This is a bit awkward...

File: FO-DICOM.Core/DicomEncoding.cs [hunk 3]
```diff
@@ -353,6 +450,43 @@ private static string DecodeFragment(byte[] fragment, int fragmentLength, Encodi
...
+                    }
+                    catch (EncoderFallbackException)
+                    {
+                        // try next encoding if any
+                    }
+                }
+            }
+
+            // the fallback uses replacement characters
+            Logger.LogWarning("Could not encode string '{Fragment}' with given encodings, " +
```
Comment (mrbean-bremen) [inline FO-DICOM.Core/DicomEncoding.cs:472]: Not sure if we have or want some strict mode that throws an exception in this case.

File: FO-DICOM.Core/DicomEncoding.cs [hunk 4]
```diff
@@ -102,6 +103,15 @@ public static void RegisterEncoding(string charset, string encoding)
...
         }
 
+        /// <summary>
+        /// Get strict encodings for given encodings.
+        /// The encodings will throw EncoderFallbackException if applied to a string they cannot encode.
+        /// </summary>
+        /// <param name="encodings">The list of non-strict encodings.</param>
+        /// <returns>Array of string encoding.</returns>
+        private static Encoding[] GetStrictEncodings(Encoding[] encodings) =>
+            (from encoding in encodings select StrictEncoding(encoding)).ToArray();
```
Comment (amoerie) [inline FO-DICOM.Core/DicomEncoding.cs]: nitpick: this might be the first use of "real LINQ" in fo-dicom. Why not just use `encodings.Select(StrictEncoding).ToArray()`?

Comment (mrbean-bremen) [reply FO-DICOM.Core/DicomEncoding.cs]: Sure, will change this!

File: FO-DICOM.Core/DicomEncoding.cs [hunk 5]
```diff
@@ -296,6 +344,56 @@ internal static string DecodeBytes(IByteBuffer buffer, Encoding[] encodings, boo
             return decodedString.ToString();
         }
 
+        internal static byte[] EncodeString(string value, Encoding[] encodings, bool isPersonName)
+        {
+            var strictEncodings = GetStrictEncodings(encodings);
```
Comment (mrbean-bremen) [inline FO-DICOM.Core/DicomEncoding.cs:337]: This part still bothers me performance-wise, as it will be called each time when serializing a string, I'll probably change that.

Comment (amoerie) [reply FO-DICOM.Core/DicomEncoding.cs:337]: Would it make sense to have an overload for the (probably common) case where there is only one encoding?

Comment (mrbean-bremen) [reply FO-DICOM.Core/DicomEncoding.cs:337]: Yes, this is one possibility - it would only involve one additional check, that should be negligible. I'll probably do that.

**Comment by mrbean-bremen** [comment]
This is ready for review, except for the release notes (which I will probably add after the release, provided that this will be merged after the release).

File: FO-DICOM.Core/DicomEncoding.cs (hunk 5 above)
Comment (mrbean-bremen) [reply FO-DICOM.Core/DicomEncoding.cs:337]: I added a simple check...

Comment (mrbean-bremen) [reply FO-DICOM.Core/DicomEncoding.cs:337]: Actually - I reverted the change. I want to be able to detect an invalid encoding also for a single encoding to issue a warning (or potentially an exception, if we want a strict mode).
At least I added a check to only handle text values (e.g. VRs which actually use the character set), not any string values.

**Comment by amoerie** [comment]
So this will stay draft until the 5.1.3 release? Just making sure I understand the reason for the "Draft" status.

**Comment by mrbean-bremen** [comment]
Well, I understood that the release should be very soon (though that was a few weeks ago), and didn't want to introduce potential regressions just before the release. Though I can of course add this to the current release (e.g. add the release notes and make it ready for review). What do you think?

**Comment by mrbean-bremen** [comment]
Given that fo-dicom.Codecs is also waiting for a release, I'd rather wait until after the release is out. This PR isn't that important - the feature has never been there, and that seems not to have been a real problem.

**Comment by mrbean-bremen** [comment]
@gofal, @amoerie - I forgot about this PR... Did you have time to check it?

**Comment by amoerie** [comment]
I already reviewed it once and the changes seemed okay back then. The tests seem very nice, so I'm not too worried about correctness.
I just hope we won't see serious performance regressions somewhere since you're touching some sensitive hot paths, but we will find out soon quickly haha. :-)

**Comment by gofal** [comment]
Looks great to me, thanks!
When doing performance checks, I noticed a bug in LazyByteBuffer, maybe you could add this to this PR. 
Usually when fo-dicom reads a file or receives a file via network, then the original binary data is kept in memory and only convertet to string if the tag is accessed. But when writing the file to disk then the original binary data is reused. So this conversion is only called, when a string-value of the DicomDataset is changed or a completely new DicomDataset is created. And in these scenarios - in order to avoid unnecessary conversion - the GetBytes is not called until it is first needed when writing the dataset to file or network.
Therefore there is the class LazyByteBuffer. This class implements IByteBuffer, but it gets a factory method, which is called implicitly when accessing the Data property.

Now the DicomWriter.OnElement method 

        public bool OnElement(DicomElement element)
        {
            WriteTagHeader(element.Tag, element.ValueRepresentation, element.Length);

            var buffer = element.Buffer;
            if (buffer is EndianByteBuffer ebb)
            {
                if (ebb.Endian != Endian.LocalMachine && ebb.Endian == _target.Endian)
                {
                    buffer = ebb.Internal;
                }
            }
            else if (_target.Endian != Endian.LocalMachine && element.ValueRepresentation.UnitSize > 1)
            {
                buffer = new SwapByteBuffer(buffer, element.ValueRepresentation.UnitSize);
            }

            WriteBuffer(_target, buffer);

            return true;
        }

accesses the element.Buffer.Data twice, once in WriteTagHeader, where the length of the data is taken, and a second time finally in WriteBuffer, where the acutal data is copied to the output stream.

And now there is the bug in LazyByte Buffer: everytime the Bytes-property is accessed, the factory method is called.

        private byte[] Bytes => _bytes();

This is why each string is encoded twice. But this can easily be solved, by caching the output of the factory method. So if you could change the Lazy ByteBuffer to

    public sealed class LazyByteBuffer : IByteBuffer
    {
        private readonly Func<byte[]> _bytes;
        // add this field
        private byte[] _byteData = null;

        public LazyByteBuffer(Func<byte[]> bytes)
        {
            _bytes = bytes ?? throw new ArgumentNullException(nameof(bytes));
        }

       // change this property to access the field, and only to call the factory method if the field is null and store the result of the factory method into the field.
        private byte[] Bytes => _byteData ??= _bytes();
        

I would have done this myself, but the branch is not in fo-dicom repository, but in your personal repository and I cannot push there.

**Review by gofal — APPROVED** [review]

**Comment by mrbean-bremen** [comment]
> I would have done this myself, but the branch is not in fo-dicom repository, but in your personal repository and I cannot push there.

This should not be a problem, I have done this several times. By default maintainers are allowed to edit PRs, this can be seen at the bottom of the right column of this page: "Maintainers are allowed to edit this pull request."

Though of course I can do that myself...

**Comment by mrbean-bremen** [comment]
Ok, done!
